p2p/security/security.go

```diff
@@ -0,0 +1,125 @@
+package security
+
+import (
+	"fmt"
+	"sync"
+
+	libp2p_network "github.com/libp2p/go-libp2p-core/network"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/pkg/errors"
+)
+
+type Security interface {
+	OnConnectCheck(net libp2p_network.Network, conn libp2p_network.Conn) error
+	OnDisconnectCheck(conn libp2p_network.Conn) error
+}
+
+type Manager struct {
+	maxConnPerIP int
+
+	mutex sync.Mutex
+	peers sync.Map // All the connected nodes, key is the Peer's IP, value is the peer's ID array
+}
+
+func NewManager(maxConnPerIP int) *Manager {
+	return &Manager{
+		maxConnPerIP: maxConnPerIP,
+	}
+}
+
+func (m *Manager) OnConnectCheck(net libp2p_network.Network, conn libp2p_network.Conn) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	ip, err := getIP(conn)
+	if err != nil {
+		return errors.Wrap(err, "failed on get ip")
+	}
+
+	value, ok := m.peers.Load(ip)
+	if !ok {
+		value = []string{}
+	}
+
+	peers, ok := value.([]string)
+	if !ok {
+		return errors.New("peers info type err")
+	}
+
+	// avoid add repeatedly
+	peerID := conn.RemotePeer().String()
+	_, ok = find(peers, peerID)
+	if !ok {
+		peers = append(peers, peerID)
+	}
+
+	if len(peers) > m.maxConnPerIP {
+		return net.ClosePeer(conn.RemotePeer())
```

Comment:
@yahtoo Could you please add a log here so we are aware when the limit has been hit, slow/blocked legitimate traffic will be hard to troubleshoot if there are no logging. cc @rlan35 @JackyWYX @LeoHChen